CI: add prerelease.yml to executes automatically some of the pre-release steps

ana opened the pull request:
This pipeline performs 5 of the 9 steps outlined `examples/testplans/release/pre.json` and helps with a sixth one "Avocado Remote Machine HTML report" - the HTML is created and saved as an artifact that can be downloaded and checked locally.
There is a detail to take into account, the containers by GitHub Actions to run the step "Avocado Test Run on RPM based installation" have modifications to make them smaller, this means it avoids saving files under `/usr/share/doc`, and avocado packaging is installing there the examples. This is why I changed the the test to run `/usr/bin/true` instead of `passtest.py`
Also as result of using a container, there is not a hostname so the HTML job report says 
`Host: 	/bin/sh: line 1: hostname: command not found"`

The three remaining tasks are:
- check all build and quality services report good indicators 
- Set the readthedocs.org token
- paginator

The two first ones can't  be automatically executed. I'm unsure about paginator, I'd need to research it.
The result of the pipeline can be browsed at https://github.com/ana/avocado/runs/3351493931

References: https://github.com/avocado-framework/avocado/issues/4601

clebergnu commented on .github/workflows/prerelease.yml:
```diff
@@ -0,0 +1,97 @@
+name: Pre-release tests
+
+on:
+  pull_request:
```
Are you suggesting we run the pre-release tests on every PR?  One problem of a pre-merge approach is that the use of [traditional COPR repos](https://copr.fedorainfracloud.org/coprs/g/avocado/avocado-latest/), for instance, would not be possible, a deployment based on the packit COPR repos would be necessary.  If that's the way to go, how can we wait for the packit build, capture its generated repository information, and inject into a workflow here?

ana replied:
> Are you suggesting we run the pre-release tests on every PR? 

No, I was using that while debugging the pipeline and forgot to update it. I was planning to run this after a manual trigger (workflow_dispatch) and weekly.

willianrampazzo commented:
> This pipeline performs 5 of the 9 steps outlined `examples/testplans/release/pre.json` and helps with a sixth one "Avocado Remote Machine HTML report" - the HTML is created and saved as an artifact that can be downloaded and checked locally.
> There is a detail to take into account, the containers by GitHub Actions to run the step "Avocado Test Run on RPM based installation" have modifications to make them smaller, this means it avoids saving files under `/usr/share/doc`, and avocado packaging is installing there the examples. This is why I changed the the test to run `/usr/bin/true` instead of `passtest.py`
> Also as result of using a container, there is not a hostname so the HTML job report says
> `Host: /bin/sh: line 1: hostname: command not found"`

What if we install the `hostname` package to make the job run?

willianrampazzo reviewed (approved):
Overall, the code looks good to me!

beraldoleal commented on .github/workflows/prerelease.yml:
```diff
@@ -0,0 +1,99 @@
+      - name: Installing Avocado in develop mode
+        run: python3 setup.py develop --user
+      - name: Run make check
+        run: make check
```
If `make check` is already executed every PR during our gating process, this makes me think about the real need of this.

I mean, if this is needed is because by some reason we accepted a PR with this failing. Yes, double-checks are good, but again, makes me think about this.

ana replied:
@beraldoleal 
Good question. I asked myself the same while adding the job. I ended adding it because it's in the checklist https://github.com/avocado-framework/avocado/blob/master/examples/testplans/release/pre.json#L10
so I thought there might be some historical reason for this. I'm fine removing it in both places in the PR and there. Thoughts ?

beraldoleal replied:
Unless we have a strong reason for keeping it, we probably could remove this. It will force us double checking before merge.

@clebergnu @willianrampazzo @richtja can you guys comment here, please?

richtja replied:
IMO we don't need this. As @beraldoleal said, is the same check form PR gating process.

willianrampazzo replied:
Looking at https://github.com/avocado-framework/avocado/actions, I don't see any checks running on the `master` branch. This means that 2 branches that succeeded in running `make check` may fail when both are merged to `master`, as one may eventually impact the other. In this case, I don't see a problem leaving the `make check` in the weekly test as it is testing master.

beraldoleal replied:
@willianrampazzo `.gitlab/workflows/ci.yml` does have a `make check` already that is executed on every PR that is trying to be part of "master".  Not sure if I got your point.

willianrampazzo replied:
Imagine I send a PR to change the software manager utility. The `make check` run on my PR succeed and is ready to be merged. Imagine someone else sends a PR at the same time I sent mine changing the software manager runner on nrunner side changing it and do not account for the changes I submitted on my PR. This PR will also succeed in running `make check`. Both PRs will be good to go because the changes on both PRs will not conflict. With both PRs merged, an issue was introduced and makes `make check` fail, but we are not running `make check` on master and the issue will appear later when someone testing its own PR find a fail on `make check` related to it.

I don't think running `make check` weekly will totally solve this problem, but at least we may see the error before someone finds then by accident. Also, I don't see one more run of `make check` per week as a problem here.

beraldoleal replied:
Thanks @willianrampazzo , of course! That is the strong reason that I was missing. :)

richtja replied:
@willianrampazzo I still do not understand to it. We are running the `make check` on master after every merge and that should catch this issue, am I right? https://github.com/avocado-framework/avocado/runs/3372112014

ana replied:
@richtja yes, but we might have some kind of "race condition" when merging commits. There is a low possibility but the check doesn't consume so much resources.

willianrampazzo replied:
@richtja, my comment was because of the removal of the Travis jobs that ran on master. Indeed, in the link you mentioned, I see they are running on the master, but for non-x86 architectures. If that is enough, I don't see a problem removing the `make check` here.

richtja replied:
I don't have a problem to have `make check` here, I just wanted to mention that we are testing the master branch and IMO that is enough.

beraldoleal commented on .github/workflows/prerelease.yml:
```diff
@@ -0,0 +1,99 @@
+      - name: Run make check
+        run: make check
+
+  avocado-pre-release-job:
```
Since we are talking about jobs here, maybe we could remove the suffix "-job" on all of those.

ana replied:
ok, I'll remove it 😄

beraldoleal commented on .github/workflows/prerelease.yml:
```diff
@@ -0,0 +1,99 @@
+          dnf -y install git
+          dnf -y module disable avocado
+          dnf -y install 'dnf-command(copr)'
+          dnf -y copr enable @avocado/avocado-latest
```
Out of curiosity: why sudo is not needed here, but needed on lines 38-39?

ana replied:
@beraldoleal Here it's running inside of the` fedora:34` container and on lines 38-39 it's running inside the ubuntu VM as regular user so you need sudo to install avocado system-wide.

beraldoleal commented on .github/workflows/prerelease.yml:
```diff
@@ -0,0 +1,99 @@
+        uses: actions/upload-artifact@v2
+        with:
+          name: avocado-html
+          path: /tmp/artifact
```
By reading your cover-letter, iiuc, the idea of this artifact is for doing a manual check after the job finishes. Is this really needed or could we automate this? I mean, I have a feeling that we are not going to look at this artifact at all.

ana replied:
I'm trying to make things easier for this step https://github.com/avocado-framework/avocado/blob/master/examples/testplans/release/pre.json#L28
If you were ignoring this step in the pre-release checks, then we can indeed not to store the artifact.

beraldoleal replied:
We are not ignoring it, my point is that maybe this is the right time to automate this check:

> "Verify that all the links such as `job-YYYY-MM-DD...` (under `Results Dir`), `1-/**/avocado/tests/**/examples/tests/passtest.py:PassTest.test` (under `Test ID`) and `debug.log` point to valid locations."

And if we keep that we way, we have to change the pre release testplan to make sure that we have instructions there, saying: "Double check the HTML artifact generated at job X.."

richtja replied:
I think we will still need some visual check on the HTML report to be sure everything is right and that the formatting is working properly.

willianrampazzo replied:
Indeed, unless we have a proper selenium-like automated test to check the HTML, we still need to do some visual checks.

beraldoleal replied:
IMO, a simple regex finding some patterns there could be enough. But I agree that we can delivery this in a different PR. But would be nice to have at least the testplan changed.

ana replied:
ok, we can have some automatic checks indeed (and I wanted to look at that eventually).

beraldoleal reviewed (commented):
Hi @ana, thanks for this. I left a few comments for you.

ana commented:
> What if we install the `hostname` package to make the job run?

The job runs, it's the resulting HTML that contains this message instead of a hostname. It's ugly but I don't find it very annoying.

beraldoleal commented:
Hi @ana, another thing that I'm missing here, is that we could change the `examples/testplans/release/pre.json` instructions and point to this workflow the specific steps, since we are going to have it automatically now.

willianrampazzo commented:
> > What if we install the `hostname` package to make the job run?
> 
> The job runs, it's the resulting HTML that contains this message instead of a hostname. It's ugly but I don't find it very annoying.

Okay, got it, no problem then.

ana commented:
> Hi @ana, another thing that I'm missing here, is that we could change the `examples/testplans/release/pre.json` instructions and point to this workflow the specific steps, since we are going to have it automatically now.

I was planning to do that in a second PR, because I still wanted to research how to make easier in the remaining steps. However, I'm thinking now it's worth it to invest more time on this. 
Thoughts: should I update this with the requested changes and update pre.json and consider it done because the longest test are done, or I should try to see how much more can be executed automatically?

beraldoleal commented:
> > Hi @ana, another thing that I'm missing here, is that we could change the `examples/testplans/release/pre.json` instructions and point to this workflow the specific steps, since we are going to have it automatically now.
> 
> @beraldoleal I was planning to do that in a second PR, because I still wanted to research how to make easier in the remaining steps. However, I'm thinking now it's worth it to invest more time on this.
> Thoughts: should I update this with the requested changes and update pre.json and consider it done because the longest test are done, or I should try to see how much more can be executed automatically?

IMO, this is already a big improvement. That can be useful for this release (today or next week). But at least testplan should be updated accordingly. And we let for a second PR more automation on top of this one, removing as much as possible the manual steps, as described here #3456. Makes sense?

ana commented:
> IMO, this is already a big improvement. That can be useful for this release (today or next week). But at least testplan should be updated accordingly. And we let for a second PR more automation on top of this one, removing as much as possible the manual steps, as described here #3456. Makes sense?

@beraldoleal 
Yes, it makes sense. I'm making a note in issue https://github.com/avocado-framework/avocado/issues/4601 and consider this done. What's left that could be run automatically is going to take way too much time to be automatized to be worth it.

I have updated this PR and added a second commit, please, take a look!

beraldoleal reviewed (approved):
4 are down! \o/ Thanks @ana.  LGTM.

clebergnu reviewed (approved):
LGTM, thanks!
